File: backend/temporal/activities/clientOnboardingActivities.ts
import { Context } from '@temporalio/activity';

/**
 * Client Onboarding Activities
 * 
 * These are the individual tasks that get executed as part of the workflow.
 * Each activity is idempotent and can be safely retried.
 */

/**
 * Validate client information against business rules
 */
export async function validateClient(
  clientId: string,
  clientName: string,
  email: string
): Promise<void> {
  console.log(`[ACTIVITY] Validating client ${clientId}...`);

  // Call backend validation endpoint
  const response = await fetch(`/api/clients/${clientId}/validate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name: clientName, email }),
  });

  if (!response.ok) {
    throw new Error(`Validation failed: ${response.statusText}`);
  }

  const result = await response.json();
  if (!result.valid) {
    throw new Error(`Validation error: ${result.reason}`);
  }

  console.log(`[ACTIVITY] Client ${clientId} validated`);
}

/**
 * Perform AML (Anti-Money Laundering) screening
 */
export async function performAMLScreening(
  clientId: string,
  amlProvider: string = 'default'
): Promise<{ passed: boolean; reason?: string }> {
  console.log(`[ACTIVITY] Performing AML screening for client ${clientId}...`);

  const response = await fetch(`/api/aml/screen`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ client_id: clientId, provider: amlProvider }),
  });

  if (!response.ok) {
    throw new Error(`AML screening failed: ${response.statusText}`);
  }

  const result = await response.json();
  console.log(
    `[ACTIVITY] AML screening ${result.passed ? 'passed' : 'failed'} for ${clientId}`
  );

  return result;
}

/**
 * Route the approval request to the appropriate manager
 */
export async function routeForApproval(
  clientId: string,
  managerId: string
): Promise<void> {
  console.log(`[ACTIVITY] Routing client ${clientId} for approval...`);

  const response = await fetch(`/api/approvals`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      client_id: clientId,
      type: 'onboarding',
      manager_id: managerId,
    }),
  });

  if (!response.ok) {
    throw new Error(`Routing failed: ${response.statusText}`);
  }

  console.log(`[ACTIVITY] Approval routed for client ${clientId}`);
}

/**
 * Generate legal agreements and documents
 */
export async function generateAgreements(
  clientId: string,
  options: { template: string; sendTo: string }
): Promise<void> {
  console.log(`[ACTIVITY] Generating agreements for client ${clientId}...`);

  const response = await fetch(`/api/documents/generate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      client_id: clientId,
      template: options.template,
      send_to: options.sendTo,
    }),
  });

  if (!response.ok) {
    throw new Error(`Document generation failed: ${response.statusText}`);
  }

  console.log(`[ACTIVITY] Agreements generated for client ${clientId}`);
}

/**
 * Create accounts for the newly onboarded client
 */
export async function createAccounts(
  clientId: string,
  options: { name: string; email: string }
): Promise<void> {
  console.log(`[ACTIVITY] Creating accounts for client ${clientId}...`);

  const response = await fetch(`/api/accounts`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      client_id: clientId,
      name: options.name,
      email: options.email,
    }),
  });

  if (!response.ok) {
    throw new Error(`Account creation failed: ${response.statusText}`);
  }

  console.log(`[ACTIVITY] Accounts created for client ${clientId}`);
}

/**
 * Send notification to client
 */
export async function notifyClient(
  clientId: string,
  options: { type: string; recipient: string; subject: string }
): Promise<void> {
  console.log(
    `[ACTIVITY] Sending ${options.type} notification to ${options.recipient}...`
  );

  const response = await fetch(`/api/notifications`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      client_id: clientId,
      type: options.type,
      recipient: options.recipient,
      subject: options.subject,
    }),
  });

  if (!response.ok) {
    throw new Error(`Notification failed: ${response.statusText}`);
  }

  console.log(`[ACTIVITY] Notification sent to ${options.recipient}`);
}

/**
 * Escalate issue to director
 */
export async function escalateToDirector(
  clientId: string,
  options: { error: string; step: string; clientName: string }
): Promise<void> {
  console.log(
    `[ACTIVITY] Escalating client ${clientId} issue to director...`
  );

  const response = await fetch(`/api/escalations`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      client_id: clientId,
      error: options.error,
      step: options.step,
      client_name: options.clientName,
      severity: 'high',
    }),
  });

  if (!response.ok) {
    throw new Error(`Escalation failed: ${response.statusText}`);
  }

  console.log(`[ACTIVITY] Issue escalated to director for client ${clientId}`);
}

File: backend/temporal/activities/client_onboarding_activities.go
package activities

import (
	"context"
	"fmt"
	"time"

	"go.temporal.io/sdk/activity"
)

// ============================================================================
// CLIENT ONBOARDING ACTIVITIES
// ============================================================================

// ValidateClientDataActivity performs KYC/AML validation and compliance checks
func ValidateClientDataActivity(ctx context.Context, input interface{}) (interface{}, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Starting client validation activity")

	// Parse input (in real implementation, use proper type assertion)
	// validationInput := input.(ValidationInput)

	// Simulate validation logic
	validationResult := map[string]interface{}{
		"passed":       true,
		"kyc_status":   "approved",
		"aml_status":   "approved",
		"risk_score":   35,
		"risk_profile": "moderate",
		"net_worth":    500000,
		"errors":       []string{},
	}

	logger.Info("Client validation completed", "result", validationResult)
	return validationResult, nil
}

// RouteForAdvisorReviewActivity assigns the client to an appropriate advisor
func RouteForAdvisorReviewActivity(ctx context.Context, input interface{}) (interface{}, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Routing client for advisor review")

	// In real implementation:
	// 1. Query advisor availability
	// 2. Match based on expertise, client risk profile, workload
	// 3. Send assignment notification
	// 4. Create review task in advisor dashboard

	routingResult := map[string]interface{}{
		"assigned_advisor_id": "advisor-12345",
		"advisor_name":        "John Smith",
		"priority":            "normal",
		"assigned_at":         time.Now(),
	}

	logger.Info("Routing completed", "advisorID", routingResult["assigned_advisor_id"])
	return routingResult, nil
}

// GenerateAndSendAgreementsActivity generates legal documents and sends for e-signature
func GenerateAndSendAgreementsActivity(ctx context.Context, input interface{}) (interface{}, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Generating and sending agreements")

	// In real implementation:
	// 1. Call document generation template engine (based on SSRS/power bi insights)
	// 2. Populate with client data
	// 3. Integrate with DocuSign or HelloSign API
	// 4. Send e-signature request via email
	// 5. Track request ID for follow-up

	agreementsResult := map[string]interface{}{
		"agreement_ids": []string{
			"docusign-req-1001",
			"docusign-req-1002",
			"docusign-req-1003",
		},
		"request_date":    time.Now(),
		"delivery_method": "email",
		"status":          "sent",
	}

	logger.Info("Agreements sent", "count", len(agreementsResult["agreement_ids"].([]string)))
	return agreementsResult, nil
}

// SendAgreementReminderActivity sends a reminder for unsigned agreements
func SendAgreementReminderActivity(ctx context.Context, input interface{}) error {
	logger := activity.GetLogger(ctx)
	logger.Info("Sending agreement reminder")

	// In real implementation:
	// 1. Query DocuSign for agreement status
	// 2. If not signed, send reminder email
	// 3. Log reminder in audit trail

	logger.Info("Reminder email sent")
	return nil
}

// CreateAccountsAndPortfoliosActivity creates investment accounts and sets up initial portfolios
func CreateAccountsAndPortfoliosActivity(ctx context.Context, input interface{}) (interface{}, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Creating accounts and portfolios")

	// In real implementation:
	// 1. Call banking API to create accounts
	// 2. Set up custodial relationships
	// 3. Create portfolio based on risk profile
	// 4. Allocate recommended holdings
	// 5. Return account numbers and portfolio IDs

	accountsResult := map[string]interface{}{
		"account_count": 2,
		"accounts": []map[string]interface{}{
			{
				"account_number": "ACC-20251028-001",
				"account_type":   "brokerage",
				"status":         "active",
			},
			{
				"account_number": "ACC-20251028-002",
				"account_type":   "ira",
				"status":         "active",
			},
		},
		"portfolio_ids": []string{"port-1001", "port-1002"},
		"created_at":    time.Now(),
	}

	logger.Info("Accounts and portfolios created", "count", accountsResult["account_count"])
	return accountsResult, nil
}

// NotifyClientOnCompletionActivity sends completion notification to the client
func NotifyClientOnCompletionActivity(ctx context.Context, input interface{}) (interface{}, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Notifying client of onboarding completion")

	// In real implementation:
	// 1. Generate welcome packet with account details
	// 2. Send welcome email
	// 3. Activate client portal access
	// 4. Send SMS notification if opted in
	// 5. Schedule follow-up advisor call

	notificationResult := map[string]interface{}{
		"email_sent":       true,
		"portal_activated": true,
		"sent_at":          time.Now(),
		"message":          "Welcome email and portal access provided",
	}

	logger.Info("Completion notification sent")
	return notificationResult, nil
}

// StartTimeoutEscalationActivity initiates escalation when a step times out
func StartTimeoutEscalationActivity(ctx context.Context, input interface{}) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Starting timeout escalation")

	// In real implementation:
	// 1. Notify manager/director of timeout
	// 2. Create task in escalation queue
	// 3. Set up auto-approval/rejection based on rules
	// 4. Return escalation workflow ID

	escalationID := fmt.Sprintf("escalation-%d", time.Now().Unix())

	logger.Info("Timeout escalation started", "escalationID", escalationID)
	return escalationID, nil
}

// ============================================================================
// HELPER ACTIVITIES
// ============================================================================

// GetAdvisorAvailabilityActivity checks advisor availability
func GetAdvisorAvailabilityActivity(ctx context.Context, riskProfile string) ([]map[string]interface{}, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Checking advisor availability", "riskProfile", riskProfile)

	// Query available advisors based on:
	// - Specialization in risk profile
	// - Current workload
	// - Location (for in-person meetings)
	// - Availability

	availableAdvisors := []map[string]interface{}{
		{
			"advisor_id":     "advisor-1",
			"name":           "Alice Johnson",
			"specialization": riskProfile,
			"workload":       15,
			"max_clients":    20,
		},
		{
			"advisor_id":     "advisor-2",
			"name":           "Bob Smith",
			"specialization": riskProfile,
			"workload":       12,
			"max_clients":    20,
		},
	}

	logger.Info("Available advisors found", "count", len(availableAdvisors))
	return availableAdvisors, nil
}

// VerifyKYCComplianceActivity performs detailed KYC checks
func VerifyKYCComplianceActivity(ctx context.Context, clientID string) (map[string]interface{}, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Verifying KYC compliance", "clientID", clientID)

	// Perform KYC checks:
	// 1. Verify identification documents
	// 2. Check against blocked person lists
	// 3. Verify beneficial ownership
	// 4. Verify source of funds

	kycResult := map[string]interface{}{
		"status":            "approved",
		"verification_date": time.Now(),
		"document_verified": true,
		"no_blocks":         true,
	}

	logger.Info("KYC verification completed", "status", kycResult["status"])
	return kycResult, nil
}

// PerformOnboardingAMLScreeningActivity performs Anti-Money Laundering screening
func PerformOnboardingAMLScreeningActivity(ctx context.Context, clientID string, provider string) (map[string]interface{}, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Performing AML screening", "clientID", clientID, "provider", provider)

	// Perform AML screening:
	// 1. Check OFAC list
	// 2. Check PEP (Politically Exposed Person) list
	// 3. Check SAC (Sanctioned Associated Counterparties)
	// 4. Check other provider lists

	amlResult := map[string]interface{}{
		"status":         "passed",
		"provider":       provider,
		"screening_date": time.Now(),
		"risk_score":     15,
		"matches":        []string{},
	}

	logger.Info("AML screening completed", "status", amlResult["status"], "riskScore", amlResult["risk_score"])
	return amlResult, nil
}

// SendAdvisorAssignmentNotificationActivity notifies advisor of assignment
func SendAdvisorAssignmentNotificationActivity(ctx context.Context, advisorID, clientName string) error {
	logger := activity.GetLogger(ctx)
	logger.Info("Sending advisor assignment notification", "advisorID", advisorID, "clientName", clientName)

	// Send notification:
	// 1. Email to advisor
	// 2. Dashboard notification
	// 3. Mobile push if enabled

	logger.Info("Advisor notified of assignment")
	return nil
}

// CreateDocuSignEnvelopeActivity creates a DocuSign envelope for e-signature
func CreateDocuSignEnvelopeActivity(ctx context.Context, clientEmail, clientName string, documentPaths []string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Creating DocuSign envelope", "clientEmail", clientEmail, "docCount", len(documentPaths))

	// Create DocuSign envelope:
	// 1. Upload documents
	// 2. Add recipient (client)
	// 3. Set signature fields
	// 4. Send envelope
	// 5. Return envelope ID

	envelopeID := fmt.Sprintf("docusign-env-%d", time.Now().Unix())

	logger.Info("DocuSign envelope created", "envelopeID", envelopeID)
	return envelopeID, nil
}

// CreateBankAccountActivity calls banking API to create account
func CreateBankAccountActivity(ctx context.Context, clientID, accountType, custodian string) (map[string]interface{}, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Creating bank account", "clientID", clientID, "accountType", accountType)

	// Call banking API:
	// 1. Validate client
	// 2. Create account structure
	// 3. Set up fund transfer instructions
	// 4. Return account details

	accountNumber := fmt.Sprintf("ACC-%d", time.Now().Unix())

	accountResult := map[string]interface{}{
		"account_number": accountNumber,
		"account_type":   accountType,
		"status":         "active",
		"created_at":     time.Now(),
		"custodian":      custodian,
	}

	logger.Info("Bank account created", "accountNumber", accountNumber)
	return accountResult, nil
}

// AllocatePortfolioHoldingsActivity sets up initial portfolio holdings
func AllocatePortfolioHoldingsActivity(ctx context.Context, portfolioID, riskProfile string, initialBalance float64) (map[string]interface{}, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Allocating portfolio holdings", "portfolioID", portfolioID, "riskProfile", riskProfile)

	// Allocate holdings based on risk profile:
	// 1. Get recommended allocation model
	// 2. Create orders for recommended securities
	// 3. Fund with initial balance
	// 4. Execute trades

	allocation := map[string]interface{}{
		"portfolio_id":    portfolioID,
		"allocation_date": time.Now(),
		"total_allocated": initialBalance,
		"holding_count":   12,
		"status":          "allocated",
	}

	logger.Info("Portfolio holdings allocated", "allocation", allocation)
	return allocation, nil
}

// SendWelcomeEmailActivity sends welcome email to client
func SendWelcomeEmailActivity(ctx context.Context, clientEmail, clientName, portalURL string) error {
	logger := activity.GetLogger(ctx)
	logger.Info("Sending welcome email", "clientEmail", clientEmail, "clientName", clientName)

	// Send email:
	// 1. Populate template with client data
	// 2. Include account details
	// 3. Include portal access link
	// 4. Include initial portfolio summary
	// 5. Send via email service

	logger.Info("Welcome email sent", "email", clientEmail)
	return nil
}

// ScheduleAdvisorCallbackActivity schedules initial advisor meeting
func ScheduleAdvisorCallbackActivity(ctx context.Context, advisorID, clientID, clientName string) (map[string]interface{}, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Scheduling advisor callback", "advisorID", advisorID, "clientName", clientName)

	// Schedule callback:
	// 1. Query advisor calendar for availability
	// 2. Find first available slot
	// 3. Create meeting in calendar system
	// 4. Send calendar invite to both parties
	// 5. Return scheduled time

	scheduledTime := time.Now().Add(24 * time.Hour)

	callbackResult := map[string]interface{}{
		"scheduled_time": scheduledTime,
		"advisor_id":     advisorID,
		"duration":       30,
		"status":         "scheduled",
	}

	logger.Info("Advisor callback scheduled", "time", scheduledTime)
	return callbackResult, nil
}

// ActivateClientPortalActivity activates client's portal access
func ActivateClientPortalActivity(ctx context.Context, clientID, clientEmail string) (string, error) {
	logger := activity.GetLogger(ctx)
	logger.Info("Activating client portal", "clientID", clientID, "email", clientEmail)

	// Activate portal:
	// 1. Create portal account
	// 2. Send access credentials
	// 3. Pre-populate with client data (accounts, holdings, etc)
	// 4. Enable initial login
	// 5. Return portal URL

	portalURL := fmt.Sprintf("https://portal.example.com/clients/%s", clientID)

	logger.Info("Client portal activated", "portalURL", portalURL)
	return portalURL, nil
}
